pkg/plumbing/spark/spark.go

```diff
@@ -0,0 +1,230 @@
+	"fmt"
+	"io"
+	"io/fs"
+	"net"
+	"path/filepath"
+
+	"github.com/google/uuid"
+	ipld "github.com/ipld/go-ipld-prime"
+	"github.com/serum-errors/go-serum"
+
+	"github.com/warptools/warpforge/pkg/dab"
+	"github.com/warptools/warpforge/pkg/logging"
+	"github.com/warptools/warpforge/pkg/plumbing/watch"
+	"github.com/warptools/warpforge/pkg/workspace"
+	"github.com/warptools/warpforge/pkg/workspaceapi"
+	"github.com/warptools/warpforge/wfapi"
+)
+
+type Config struct {
+	Fsys             fs.FS     // Root filesystem. Generally os.DirFS("/").
+	ModulePath       string    // Path to module. Will attempt to find closest module within the workspace.
+	WorkingDirectory string    // Absolute path to working directory
+	Dialer           Dialer    // Method of dialing the server. May be nil for default.
+	OutputMarkup     string    // Output Markup type
+	OutputStyle      string    // Output style
+	OutputColor      bool      // True for colored output
+	OutputStream     io.Writer // where output should go
+}
+
+// ECodeSpark* error codes are sentinels used for spark output.
+const (
+	ECodeSparkNoModule    = "warpforge-spark-no-module"    // locally can't find module path
+	ECodeSparkNoWorkspace = "warpforge-spark-no-workspace" // locally can't find workspace
+	ECodeSparkNoSocket    = "warpforge-spark-no-socket"    // locally can't find socket or can't dial the socket
+	ECodeSparkInternal    = "warpforge-spark-internal"     // other errors, including serialization, broken comms, invalid data errors
+	ECodeSparkServer      = "warpforge-spark-server"       // server error response
+)
+
+// searchPath will canonicalize the ModulePath argument.
+// May panic if WorkingDirectory is not an absolute path.
+func (c *Config) searchPath() string {
+	if filepath.IsAbs(c.ModulePath) {
+		return filepath.Clean(c.ModulePath)
+	}
+	if !filepath.IsAbs(c.WorkingDirectory) {
+		panic("working directory must be an absolute path")
+	}
+	return filepath.Join(c.WorkingDirectory, c.ModulePath)
+}
+
+func (c *Config) run(ctx context.Context) (workspaceapi.ModuleStatusAnswer, error) {
+	var empty workspaceapi.ModuleStatusAnswer
+	logger := logging.Ctx(ctx)
+
+	searchPath := c.searchPath()
+	logger.Debug("", "search path: %q", searchPath)
+
+	ws, _, err := workspace.FindWorkspace(c.Fsys, "", searchPath)
+	if err != nil {
+		return empty, serum.Error(ECodeSparkNoWorkspace, serum.WithCause(err))
+	}
+	if ws == nil {
+		return empty, serum.Error(ECodeSparkNoWorkspace, serum.WithMessageLiteral("workspace not found"))
+	}
+	_, wsPath := ws.Path()
+	logger.Debug("", "ws path: %q", wsPath)
+
+	logger.Debug("", "module search path: %q", searchPath)
+	modulePath, _, err := dab.FindModule(c.Fsys, wsPath, searchPath)
+	if err != nil {
+		return empty, serum.Error(ECodeSparkNoModule, serum.WithCause(err))
+	}
+
+	query := workspaceapi.ModuleStatusQuery{
+		Path:          modulePath,
+		InterestLevel: workspaceapi.ModuleInterestLevel_Query,
+	}
+	logger.Info("", "ModulePath: %s", query.Path)
```

Comment:
This should likely be a `logger.Debug`, currently it appears by default when running spark.

Reply:
Also, missing leading `/`